eventauth/eventcontent.go

```diff
@@ -0,0 +1,280 @@
+	banLevel          int64
+	inviteLevel       int64
+	kickLevel         int64
+	redactLevel       int64
+	userLevels        map[string]int64
+	userDefaultLevel  int64
+	eventLevels       map[string]int64
+	eventDefaultLevel int64
+	stateDefaultLevel int64
+}
+
+// userLevel returns the power level a user has in the room.
+func (c *powerLevelContent) userLevel(userID string) int64 {
+	level, ok := c.userLevels[userID]
+	if ok {
+		return level
+	}
+	return c.userDefaultLevel
+}
+
+// eventLevel returns the power level needed to send an event in the room.
+func (c *powerLevelContent) eventLevel(eventType string, eventStateKey *string) int64 {
+	if eventType == "m.room.third_party_invite" {
+		// Special case third_party_invite events to have the same level as
+		// m.room.member invite events.
+		// https://github.com/matrix-org/synapse/blob/v0.18.5/synapse/api/auth.py#L182
+		return c.inviteLevel
+	}
+	level, ok := c.eventLevels[eventType]
+	if ok {
+		return level
+	}
+	if eventStateKey != nil {
+		return c.stateDefaultLevel
+	}
+	return c.eventDefaultLevel
+}
+
+// load the power level content from the power level event in the auth events.
+func (c *powerLevelContent) load(authEvents AuthEvents, creatorUserID string) error {
+	powerLevelsEvent, err := authEvents.PowerLevels()
+	if err != nil {
+		return err
+	}
+	if powerLevelsEvent != nil {
+		return c.parse(*powerLevelsEvent)
+	}
+
+	// If there are no power leves then fall back to defaults.
+	c.defaults()
+	// If there is no power level event then the creator gets level 100
+	// https://github.com/matrix-org/synapse/blob/v0.18.5/synapse/api/auth.py#L569
+	c.userLevels = map[string]int64{creatorUserID: 100}
+	return nil
+}
+
+// defaults sets the power levels to their default values.
+func (c *powerLevelContent) defaults() {
+	// Default invite level is 0.
+	// https://github.com/matrix-org/synapse/blob/v0.18.5/synapse/api/auth.py#L426
+	c.inviteLevel = 0
+	// Default ban, kick and redacts levels are 50
+	// https://github.com/matrix-org/synapse/blob/v0.18.5/synapse/api/auth.py#L376
+	// https://github.com/matrix-org/synapse/blob/v0.18.5/synapse/api/auth.py#L456
+	// https://github.com/matrix-org/synapse/blob/v0.18.5/synapse/api/auth.py#L1041
+	c.banLevel = 50
+	c.kickLevel = 50
+	c.redactLevel = 50
+	// Default user level is 0
+	// https://github.com/matrix-org/synapse/blob/v0.18.5/synapse/api/auth.py#L558
+	c.userDefaultLevel = 0
+	// Default event level is 0, Default state level is 50
+	// https://github.com/matrix-org/synapse/blob/v0.18.5/synapse/api/auth.py#L987
+	// https://github.com/matrix-org/synapse/blob/v0.18.5/synapse/api/auth.py#L991
+	c.eventDefaultLevel = 0
+	c.stateDefaultLevel = 50
+
+}
+
+// parse the power level content from an event.
+func (c *powerLevelContent) parse(event Event) error {
+	// Set the levels to their default values.
+	c.defaults()
+
+	// We can't extract the JSON directly to the powerLevelContent because we
+	// need to convert string values to int values.
```

Comment:
This should probably also be mentioned in the struct definition. I nearly whined at you for it.